htlcswitch/link.go

```diff
@@ -2503,12 +2610,59 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
 	}, false)
 }
 
-// fail helper function which is used to encapsulate the action necessary for
-// proper disconnect.
-func (l *channelLink) fail(format string, a ...interface{}) {
+// fail is a function which is used to encapsulate the action necessary for
+// properly failing the link. It takes a linkFailureError, which will be used
+// to determine if we should force close the channel, and if we should send an
+// error message to the remote peer. If we decide to force close, we will in
+// addition to disconnecting the peer unilateraly close the channel.
+func (l *channelLink) fail(linkErr linkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+
+	// We launch a go routine to disconnect the peer and optionally force
+	// close the channel. We must do this since these call will attempt to
+	// stop the link, and wait for it to finish shutting down, and we must
+	// return from fail() to allow the link to shut down completely.
+	go func() {
+		// Send an error to the peer, why we failed the channel.
```

Comment:
The reason that is hard to do without starting a goroutine is that we must allow the `fail()` call to return for the link to shutdown (to allow the `Stop()` call triggered by the `Disconnect` to return).

We could add a call to `WaitForDisconnected` here, but I don't think it would really matter (the call to `ForceClose` would stop and wait for the disconnect to finish regardless)

Reply:
I'd be in favor of synchronizing this more heavily. Even if the disconnect goes through, there could still be messages queued in the mailbox that could be processed in some instances. There are places in the link where we disconnect, but don't instruct the htlc manager to exit, which seems like it could be a problem.

One way we can go about this is to use `fail` to _schedule_ a disconnect and/or force close by caching the `linkFailureError` inside the link, and then executing the bulk of this logic in a defer statement as the `htlcManager` is actually exiting.

One thing to consider, is that even after some calls to `fail` complete, the main event loop can still execute if we have messages queued up in mailboxes, event timers/block epochs fire, etc. We can prevent some of these from being delivered by nilling out channels such as `downstream` and `upstream`, stopping the packet overflow queue, etc.

Downside of this is that it could end up being a modest reworking of the link shutdown :(

Reply:
What about adding a buffered `linkFailures chan linkFailureError` to the link, that we can send on from `fail(). We can then check this channel in the `htlcManager` before it exits.

This wouldn't help with the deadlock scenario though, as we would still want the link to properly exit before we can execute the force close. What we could do to resolve this is to define a failure callback in the link config (or a channel if we don't want to introduce a new potential deadlock scenario), that is populated by the switch. We could then trigger the force close (and the disconnect) within this callback after the link is properly shut down.

Thoughts?

Reply:
If the `linkFailures` channel is being read in the main event loop, there's no guarantee that the failure will be pulled off next, which may allow other updates/actions to continue before actually exiting.

IMO the easiest solutions is just to make a `linkFailure` member variable on the channel link, and just check at the beginning of each loop to see if the value is non-nil. This way, we always exit immediately after detecting that a `case` in the main event loop failed. Most of the logic currently in `fail` could be moved to the end of `htlcManager`, i.e. sending errors/force closing, to handle the failure gracefully.

We may also need to switch the ordering of force closing and disconnection, otherwise we could disconnect, and then reconnect quickly and the channel will still be open in channeldb. 

After we've broken out of the main event loop and force closed, it seems the synchronicity requirements for disconnect become less stringent. It should be safe just to do `go Disconnect()` at the end of the `htlcManager`, i think?

Reply:
Added a memeber variable as suggested, to make sure we don't process more updates after a failure is encountered.

In addition I moved the actual force close the `ChannelLinkConfig` creator (`peer`), since it has better control of the link lifecycle.

Finally, as suggested I made disconnection be done after force close is completed.

PTAL.

Reply:
@halseth sweet, taking another look :)